File: python/packages/ai/teams/ai/models/azure_openai/client/azure_openai_client.py
"""
Copyright (c) Microsoft Corporation. All rights reserved.
Licensed under the MIT License.
"""

from typing import Dict, List, Literal, Union

from aiohttp import ClientResponseError, ClientSession

from teams.ai.models.openai import (
    CreateModerationResponse,
    OpenAIClient,
    OpenAIClientError,
    OpenAIClientResponse,
)


class AzureOpenAIClient(OpenAIClient):
    """
    Azure OpenAI Http Client
    """

    _api_version: str

    @property
    def _headers(self) -> Dict[str, str]:
        headers = {
            "Content-Type": "application/json",
            "User-Agent": "Microsoft Teams Conversational AI SDK",
            "Ocp-Apim-Subscription-Key": self._api_key,
        }

        return headers

    def __init__(self, api_key: str, *, base_url: str, api_version: str = "2022-12-01") -> None:
        super().__init__(api_key, base_url=base_url)
        self._api_version = api_version

    async def create_moderation(
        self,
        input: Union[str, List[str]],
        *,
        model: Literal[
            "text-moderation-stable", "text-moderation-latest"
        ] = "text-moderation-stable",
    ) -> OpenAIClientResponse[CreateModerationResponse]:
        async with ClientSession(
            base_url=self._base_url,
            headers=self._headers,
        ) as session:
            try:
                res = await session.post(
                    f"/contentsafety/text:analyze?api-version={self._api_version}",
                    json={"text": input},
                )
                data = await res.json()

                if res.status >= 400:
                    raise ClientResponseError(
                        res.request_info, res.history, status=res.status, message=data
                    )

                return OpenAIClientResponse[CreateModerationResponse](
                    status=res.status,
                    headers=res.headers,
                    data=CreateModerationResponse.from_dict(data),
                )
            except ClientResponseError as err:
                raise OpenAIClientError(
                    status=err.status, message=err.message, headers=err.headers
                ) from err

File: js/packages/teams-ai/src/internals/AzureOpenAIClient.ts
/**
 * @module teams-ai
 */
/**
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License.
 */

import {
    CreateChatCompletionRequest,
    CreateChatCompletionResponse,
    CreateCompletionRequest,
    CreateCompletionResponse,
    CreateEmbeddingRequest,
    CreateEmbeddingResponse,
    ModerationInput,
    ModerationResponse,
    OpenAICreateChatCompletionRequest,
    OpenAICreateCompletionRequest,
    OpenAICreateEmbeddingRequest
} from './types';
import { OpenAIClient, OpenAIClientOptions, OpenAIClientResponse } from './OpenAIClient';

/**
 * @private
 */
export interface AzureOpenAIClientOptions extends OpenAIClientOptions {
    /**
     * Azure OpenAI endpoint.
     */
    endpoint: string;

    /**
     * Optional. Which Azure API version to use. Defaults to latest.
     */
    apiVersion?: string;
}

/**
 * @private
 * @class
 * @implements {OpenAIClient}
 * `AzureOpenAIClient` Allows for Azure hosted OpenAI clients to be created and used. As of 4/4/2023, access keys must be specifically assigned to be used with this client.
 */
export class AzureOpenAIClient extends OpenAIClient {
    public constructor(options: AzureOpenAIClientOptions) {
        super(options);

        // Validate endpoint
        if (!options.endpoint) {
            throw new Error(`AzureOpenAIClient initialized without an 'endpoint'.`);
        }
    }

    public createCompletion(request: CreateCompletionRequest): Promise<OpenAIClientResponse<CreateCompletionResponse>> {
        const clone = Object.assign({}, request) as OpenAICreateCompletionRequest;
        const deployment = this.removeModel(clone);
        const endpoint = (this.options as AzureOpenAIClientOptions).endpoint;
        const apiVersion = (this.options as AzureOpenAIClientOptions).apiVersion ?? '2022-12-01';
        const url = `${endpoint}/openai/deployments/${deployment}/completions?api-version=${apiVersion}`;
        return this.post(url, clone);
    }

    public createChatCompletion(
        request: CreateChatCompletionRequest
    ): Promise<OpenAIClientResponse<CreateChatCompletionResponse>> {
        const clone = Object.assign({}, request) as OpenAICreateChatCompletionRequest;
        const deployment = this.removeModel(clone);
        const endpoint = (this.options as AzureOpenAIClientOptions).endpoint;
        const apiVersion = (this.options as AzureOpenAIClientOptions).apiVersion ?? '2023-03-15-preview';
        const url = `${endpoint}/openai/deployments/${deployment}/chat/completions?api-version=${apiVersion}`;
        return this.post(url, clone);
    }

    public createEmbedding(request: CreateEmbeddingRequest): Promise<OpenAIClientResponse<CreateEmbeddingResponse>> {
        const clone = Object.assign({}, request) as OpenAICreateEmbeddingRequest;
        const deployment = this.removeModel(clone);
        const endpoint = (this.options as AzureOpenAIClientOptions).endpoint;
        const apiVersion = (this.options as AzureOpenAIClientOptions).apiVersion ?? '2022-12-01';
        const url = `${endpoint}/openai/deployments/${deployment}/embeddings?api-version=${apiVersion}`;
        return this.post(url, clone);
    }

    public createModeration(request: ModerationInput): Promise<OpenAIClientResponse<ModerationResponse>> {
        const endpoint = (this.options as AzureOpenAIClientOptions).endpoint;
        const url = `${endpoint}/contentsafety/text:analyze?api-version=${
            (this.options as AzureOpenAIClientOptions).apiVersion
        }`;
        return this.post(url, request);
    }

    protected addRequestHeaders(headers: Record<string, string>, options: OpenAIClientOptions): void {
        headers[options.headerKey ?? 'api-key'] = options.apiKey;
    }

    private removeModel(request: { model?: string }): string {
        const model = request.model;
        delete request.model;

        if (model) {
            return model;
        } else {
            return '';
        }
    }
}
